src/atomic_file/atomic.rs

```diff
@@ -0,0 +1,369 @@
+    fn write(&mut self, buf: &[u8]) -> Result<usize> {
+        (&self.file).write(buf)
+    }
+
+    fn flush(&mut self) -> Result<()> {
+        (&self.file).flush()
+    }
+}
+
+impl Drop for TmpFile {
+    fn drop(&mut self) {
+        let _ = std::fs::remove_file(&self.path);
+    }
+}
+
+#[derive(Clone)]
+pub struct ReadOnlyFile {
+    version: usize,
+    path: PathBuf,
+}
+
+/// This struct is the only way to read the file. Both path and version are private
+impl ReadOnlyFile {
+    /// Open the underlying file, which can be read from but not written to.
+    /// May return `Ok(None)`, which means that no version
+    /// of the`AtomicFile` has been created yet.
+    pub fn open(&self) -> Result<Option<File>> {
+        if self.version != 0 {
+            Ok(Some(File::open(&self.path)?))
+        } else {
+            Ok(None)
+        }
+    }
+
+    pub fn read_to_string(&self) -> Result<String> {
+        match self.open() {
+            Ok(None) => Err(Error::new(ErrorKind::NotFound, "File not found")),
+            Ok(Some(mut file)) => {
+                let mut buff = String::new();
+                file.read_to_string(&mut buff)?;
+                Ok(buff)
+            }
+            Err(e) => Err(e),
+        }
+    }
+
+    pub fn read_content(&self) -> Result<Vec<u8>> {
+        match self.open() {
+            Ok(None) => Err(Error::new(ErrorKind::NotFound, "File not found")),
+            Err(e) => Err(e),
+            Ok(Some(mut file)) => {
+                let mut buf = vec![];
+                file.read_to_end(&mut buf)?;
+                Ok(buf)
+            }
+        }
+    }
+}
+
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct AtomicFile {
+    directory: PathBuf,
+    prefix: String,
+}
+
+fn parse_version(filename: Option<&str>) -> Option<usize> {
+    let (_, version) = filename?.rsplit_once('.')?;
+    version.parse().ok()
+}
+
+impl AtomicFile {
+    pub fn new(path: impl Into<PathBuf>) -> crate::Result<Self> {
+        let directory = path.into();
+        // Should assure transfert on internet is safe before sending files containing
+        // this machine_id on the prefix
+        let machine_id = machine_uid::get()?;
+        std::fs::create_dir_all(&directory)?;
+        let filename: &str = match directory.file_name() {
+            Some(name) => name.to_str().unwrap(),
+            None => Err(std::io::Error::new(
+                std::io::ErrorKind::InvalidInput,
+                "`path` must specify a directory name",
+            ))?,
+        };
+        let prefix = format!("{}_{}.", filename, machine_id);
+        Ok(Self { directory, prefix })
+    }
+
+    /// Return the latest version together with vector of the
+    /// files matching this version. Multiple files for the same version
+    /// can appear due to usage of file syncronization. Different devices
+    /// can create same version simultaneously.
+    fn latest_version(&self) -> Result<(Vec<ReadOnlyFile>, usize)> {
```

Comment:
_(Nitpick)_ Putting version first in the returned tuple would be more readable since the name of the function is explicit about it.